File: train/verify_bake.py
"""
Hold index.html's bake against the PyTorch one, kernel by kernel.

    node train/verify_bake.mjs train/runs/add4/preset.json > train/browser_bank.json
    python3 train/verify_bake.py train/runs/add4/preset.json

A trained world is only a result if it is the SAME world in the browser. The
failure mode this guards is silent: an unrecognised field on a term is not
rejected, it is skipped, so an angular kernel loads as the ring you get by
deleting its angle -- and the preset runs, and looks fine, and is a different
simulation. Cosine similarity per channel is the number to read; anything below
0.999 means the export is not faithful.
"""

import json
import os
import sys

import numpy as np
import torch

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import fieldlife as fl

HERE = os.path.dirname(os.path.abspath(__file__))
preset = json.load(open(sys.argv[1] if len(sys.argv) > 1
                        else os.path.join(HERE, "runs/add4/preset.json")))
browser = json.load(open(os.path.join(HERE, "browser_bank.json")))

C, KR = preset["C"], browser["KR"]
print(f"{C} channels, shared stencil half-width {KR}, "
      f"reaches {[round(k['R'], 3) for k in preset['kernels'][:C]]}")

ours = fl.bake_from_config(preset["kernels"], C, KR, dtype=torch.float64).numpy()
worst = 1.0
for c in range(C):
    b = browser["banks"][c]
    k = np.array(b["w"], dtype=np.float64)
    n = int(round(len(k) ** 0.5))
    k = k.reshape(n, n)
    if n != 2 * KR + 1:                        # inset a coarser bake, as uploadKernels does
        o = KR - b["kr"]
        pad = np.zeros((2 * KR + 1, 2 * KR + 1))
        pad[o:o + n, o:o + n] = k
        k = pad
    a = ours[c]
    cos = float((a * k).sum() / max(np.linalg.norm(a) * np.linalg.norm(k), 1e-30))
    worst = min(worst, cos)
    print(f"  ch{c}  baked at kr {b['kr']:2d}   cosine {cos:+.6f}   "
          f"max |diff| {np.abs(a - k).max():.2e}")

print(f"\nworst cosine {worst:+.6f}  ->  "
      + ("the browser bakes what was trained"
         if worst > 0.999 else "THE EXPORT IS NOT FAITHFUL"))

File: train/verify_bake.mjs
/*
 * Does index.html bake the kernel the trainer thinks it does?
 *
 *   node train/verify_bake.mjs [preset.json] > train/browser_bank.json
 *
 * Bakes a preset's kernel bank inside the real page and prints it, so
 * train/verify_bake.py can hold it against the PyTorch bake. This is the check
 * the lizard experiment learned to run: a kernel field index.html does not know
 * about is not rejected, it is silently ignored, and the preset then loads as a
 * different, rotationally symmetric world with nothing anywhere reporting it.
 *
 * Only the bake is exercised -- no simulation, no WebGL -- which is why it is
 * seconds rather than minutes and safe to run beside a training job.
 */
import { chromium } from "playwright";
import { createServer } from "node:http";
import { readFile } from "node:fs/promises";
import { extname, join, resolve } from "node:path";

const ROOT = resolve(import.meta.dirname, "..");
const CHROME = "/opt/pw-browsers/chromium-1194/chrome-linux/chrome";
const PRESET = process.argv[2] ?? join(ROOT, "train/runs/add4/preset.json");
const preset = JSON.parse(await readFile(PRESET, "utf8"));

const MIME = { ".html": "text/html", ".json": "application/json", ".png": "image/png" };
const server = createServer(async (req, res) => {
  const p = join(ROOT, decodeURIComponent(req.url.split("?")[0]).replace(/^\/+/, "") || "index.html");
  try {
    res.writeHead(200, { "content-type": MIME[extname(p)] ?? "application/octet-stream" });
    res.end(await readFile(p));
  } catch { res.writeHead(404); res.end("no"); }
});
await new Promise(r => server.listen(0, "127.0.0.1", r));

const browser = await chromium.launch({
  executablePath: CHROME,
  args: ["--use-gl=angle", "--use-angle=swiftshader", "--enable-unsafe-swiftshader"],
});
const page = await browser.newPage();
page.on("pageerror", e => console.error("page error:", e.message));
await page.goto(`http://127.0.0.1:${server.address().port}/index.html`,
                { waitUntil: "domcontentloaded" });
await page.waitForFunction(() => typeof bakeKernel === "function", null, { timeout: 60000 });

const out = await page.evaluate(({ preset }) => {
  const C = preset.C, ks = preset.kernels.slice(0, C);
  // uploadKernels' own arithmetic: one shared grid from the widest reach, and
  // every other channel resampled to a coarser one in proportion to its own.
  const Rmax = Math.max(...ks.map(k => k.R), 1);
  const KR = Math.max(3, Math.min(15, Math.round(Rmax)));
  return {
    KR, Rmax,
    banks: ks.map(k => {
      const kr = Math.max(2, Math.round(KR * Math.min(1, k.R / Rmax)));
      return { kr, w: Array.from(bakeKernel(k, kr)) };
    }),
    separable: separablePlan === undefined ? null : "checked-elsewhere",
  };
}, { preset });

console.log(JSON.stringify(out));
await browser.close();
server.close();
